elan/rpc/rpc.go

```diff
@@ -571,42 +575,52 @@ func (s *server) readBlob(ctx context.Context, key string, offset, length int64)
+func (s *server) readAllBlob(ctx context.Context, prefix string, digest *pb.Digest) ([]byte, error) {
+	b, _, err := s.readAllBlobBatched(ctx, prefix, digest, false, false)
+	return b, err
+}
+
+func (s *server) readAllBlobBatched(ctx context.Context, prefix string, digest *pb.Digest, batched, allowCompression bool) ([]byte, bool, error) {
 	if len(digest.Hash) < 2 {
-		return nil, fmt.Errorf("Invalid hash: [%s]", digest.Hash)
+		return nil, false, fmt.Errorf("Invalid hash: [%s]", digest.Hash)
 	} else if s.isEmpty(digest) {
-		return nil, nil
+		return nil, false, nil
+	}
+	// TODO(peterebden): Is it worth trying to cache any knowledge of where to go first for blobs?
+	//                   Or just guessing based on blob size?
+	if allowCompression {
+		if b, err := s.readAllBlobCompressed(ctx, digest, s.compressedKey(prefix, digest, true), batched, true); err == nil {
+			return b, true, nil
+		}
 	}
+	// Try uncompressed
+	if b, err := s.readAllBlobCompressed(ctx, digest, s.compressedKey(prefix, digest, false), batched, false); err == nil || allowCompression || prefix == "ac" {
+		return b, false, err
+	}
+	// If we don't allow compression, we still have to check the compressed CAS for the client
+	b, err := s.readAllBlobCompressed(ctx, digest, s.compressedKey(prefix, digest, true), batched, true)
+	if err != nil {
+		return nil, false, err
+	}
+	b, err = s.decompressor.DecodeAll(b, make([]byte, 0, digest.SizeBytes))
+	return b, false, err
+}
+
+func (s *server) readAllBlobCompressed(ctx context.Context, digest *pb.Digest, key string, batched, compressed bool) ([]byte, error) {
 	s.limiter <- struct{}{}
 	defer func() { <-s.limiter }()
 	start := time.Now()
 	defer func() { readDurations.Observe(time.Since(start).Seconds()) }()
-	if b, err := s.bucket.ReadAll(ctx, s.compressedKey(prefix, digest, compressed)); err == nil {
-		blobsServed.WithLabelValues(batchLabel(batched, false), compressorLabel(compressed), compressorLabel(compressed)).Inc()
-		return b, nil
-	} else if prefix != "cas" {
-		return nil, handleNotFound(err, digest.Hash) // No point going on, the AC is never compressed.
-	}
-	// If we get here, it failed, so we need to check the compressed / non-compressed version.
-	b, err := s.bucket.ReadAll(ctx, s.compressedKey(prefix, digest, !compressed))
+	b, err := s.bucket.ReadAll(ctx, key)
 	if err != nil {
 		return nil, handleNotFound(err, digest.Hash)
 	}
-	// We now need to either compress or decompress the payload, since we read the other version.
-	buf := make([]byte, 0, digest.SizeBytes)
-	if compressed {
-		return s.compressor.EncodeAll(b, buf), nil
-	}
-	b, err = s.decompressor.DecodeAll(b, buf)
-	if err != nil {
-		return nil, err
-	}
-	blobsServed.WithLabelValues(batchLabel(batched, false), compressorLabel(compressed), compressorLabel(!compressed)).Inc()
+	blobsServed.WithLabelValues(batchLabel(batched, false), compressorLabel(compressed), compressorLabel(compressed)).Inc()
```

Comment:
`compressorLabel(compressed),compressorLabel(!compressed)`

Reply:
Hmmmm on some thought I've set the second one to not applicable. They're "compressor used" and "compressor preferred", in this new world there is no preferred compressor when serving batched blobs.